backend/dataall/modules/redshift_datasets_shares/services/redshift_table_share_processor.py

```diff
@@ -18,11 +33,477 @@ def __init__(self, session, share_data, shareable_items, reapply=False):
+                        )
+                        redshift_client_in_target.revoke_schema_usage_access_to_redshift_role(
+                            schema=external_schema, rs_role=self.redshift_role
+                        )
+                        if local_db_exists:
+                            redshift_client_in_target.revoke_database_usage_access_to_redshift_role(
+                                database=local_db, rs_role=self.redshift_role
+                            )
+                        else:
+                            log.info('Database does not exist, no permissions need to be revoked')
+
+                    # 6) (in target namespace) If no more tables are shared with any role in this namespace, drop local database
+                    # 7) (in source namespace) If no more tables are shared with any role in this namespace, drop datashare
+                    if (
+                        RedshiftShareRepository.count_dataset_shared_items_with_namespace(
+                            session=self.session,
+                            dataset_uri=self.dataset.datasetUri,
+                            connection_uri=self.share.principalId,
+                        )
+                        == 0
+                    ):
+                        log.info(
+                            f'No other tables of this dataset are shared with this namespace {self.target_connection.nameSpaceId}'
+                        )
+                        redshift_client_in_target.drop_database(database=local_db)
+                        redshift_client_in_source.drop_datashare(datashare=self.datashare_name)
+
+            except Exception as e:
+                log.error(
+                    f'Failed to clean up shared resources in redshift datashares for redshift dataset {self.dataset.name} '
+                    f'tables {[t.name for t in self.tables]} '
+                    f'from source {self.source_connection.name} in namespace {self.source_connection.nameSpaceId}'
+                    f'with target {self.target_connection.name} in namespace {self.target_connection.nameSpaceId}'
+                    f'due to: {e}'
+                )
+                if not self.reapply:
+                    new_state = revoked_item_SM.run_transition(ShareItemActions.Failure.value)
+                    revoked_item_SM.update_state(self.session, self.share.shareUri, new_state)
+                else:
+                    for table in self.tables:
+                        share_item = ShareObjectRepository.find_sharable_item(
+                            self.session, self.share.shareUri, table.rsTableUri
+                        )
+                        ShareStatusRepository.update_share_item_health_status(
+                            self.session, share_item, ShareItemHealthStatus.Unhealthy.value, str(e), datetime.now()
+                        )
+                return False
```

Comment:
I see this block which is why I imagine we need to be able to transition from `Revoke_Succeeded` to `Revoke_Failed` if cannot clean up the DB / Datashare

I would think either (1) only call `update_state` once at the end of the entire revoke/ clean up rather than calling once after table clean up and again after datashare clean up OR (2) we do not update to `Revoke_Failed` of the non-table resources -- but this could leave some orphaned resources / not sure how it affects re-sharing same table to same target namespace later on

Reply:
That was my dilemma. (1) is doable but trickier for the checks on the existence of other tables; I can look into the proper query that would be needed. (2) that was my thought, it might look as successful but in practice there might be leftovers. When it comes to re-sharing I made sure that it is possible even with orphan resources, but my concern is more on the orphan resources. a datashare shared with another cluster gives a lot of access rights.

In short, I will look at option (1) again and try to come up with the ultimate good solution


**by the way this issue is also present for Lake Formation glue sharing

Reply:
I think the latest changes are what we are looking for. But I want your opinion on point 4) I think it is logical, if a revoke has failed, the clean-up does not happen until every table is revoked with 1, 2, 3

```
        For each table:
            Update table status with Start Action (Revoke_Approved ---> Revoke_In_Progress)
            try:
                1) 2) 3) 
            except:
                Update table status with Failure Action (Revoke_In_Progress ---> Revoke_Failed)
        If the previous is successful, we proceed to clean-up shared resources across datashares:
        try:
            4) (in target namespace) If no more tables are shared with the redshift role, revoke usage access to the external schema to the redshift role ----> IF A TABLE IS IN REVOKE_FAILED, IT WON'T CLEAN-UP THE SHARE
            5) 6) 7)
            Update NON-FAILED tables with Success Action (Revoke_In_Progress ---> Revoke_Succeeded)
        except:
            Update tables with Failure Action (Revoke_In_Progress ---> Revoke_Failed)
```